### PR TITLE
[Internal] CTL: Adds scenario for Change feed pull

### PR DESCRIPTION
**Pull request by jenglish81**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Include samples if adding new API, and include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (ealsur) [inline .vscode/launch.json:1]: delete this file

File: .vscode/tasks.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (ealsur) [inline .vscode/tasks.json:1]: delete this file

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/.vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/.vscode/launch.json:1]: delete this file

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/.vscode/tasks.json [hunk 4]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/.vscode/tasks.json:1]: delete this file

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    using System.Diagnostics;
+
+    using App.Metrics;
+    using App.Metrics.Counter;
+    using App.Metrics.Gauge;
+
+    using Microsoft.Azure.Cosmos;
+    using Microsoft.Extensions.Logging;
+
+    using Newtonsoft.Json;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:22]: Make sure you are using all the usings, if there are any not used, please remove it

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            CosmosClient cosmosClient,
+            ILogger logger,
+            IMetrics metrics,
+            string loggingContextIdentifier,
+            CancellationToken cancellationToken)
+        {
+            Stopwatch stopWatch = new Stopwatch();
+            stopWatch.Start();
+
+            //Repeat the loop of creating a new Change Feed Iterator and all the steps until the run-time is completed.
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:63]: remove comment

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+        public async Task RunAsync(
+            CTLConfig config,
+            CosmosClient cosmosClient,
+            ILogger logger,
+            IMetrics metrics,
+            string loggingContextIdentifier,
+            CancellationToken cancellationToken)
+        {
+            Stopwatch stopWatch = new Stopwatch();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:60]: can be simplified:
```suggestion
            Stopwatch stopWatch = Stopwatch.StartNew();
```

And remove the Start line below

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+                int documentTotal = 0;
+                string continuation = null;
+                Container container = cosmosClient.GetContainer(config.Database, config.Collection);
+                FeedIterator<Dictionary<string, string>> changeFeedPull = container.GetChangeFeedIterator<Dictionary<string, string>>
+                                                                                (ChangeFeedStartFrom.Beginning(), ChangeFeedMode.Incremental);
+                try 
+                {
+                    while (changeFeedPull.HasMoreResults)
+                    {
+                        //Call ReadNextAsync on the iterator
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:75]: Remove comment

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                        FeedResponse<Dictionary<string, string>> response = await changeFeedPull.ReadNextAsync();
+                        //If there are new documents in the response, count them in a variable
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:77]: Remove comment

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                        documentTotal += response.Count;
+                        continuation = response.ContinuationToken;
+                        //If there are no new changes, stop/break out of the iterator.
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:80]: Remove comment

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+            ILogger logger,
+            IMetrics metrics,
+            string loggingContextIdentifier,
+            CancellationToken cancellationToken)
+        {
+            Stopwatch stopWatch = new Stopwatch();
+            stopWatch.Start();
+
+            //Repeat the loop of creating a new Change Feed Iterator and all the steps until the run-time is completed.
+            while(stopWatch.Elapsed <= config.RunningTimeDurationAsTimespan)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:64]: indenting and spacing. If you open this in VS Code, I think you can use the Format File command. If you open it in VS, saving the file or adding new lines might do the trick.

For example, there needs to be an space between `while` and `(`

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+                        //If there are new documents in the response, count them in a variable
+                        documentTotal += response.Count;
+                        continuation = response.ContinuationToken;
+                        //If there are no new changes, stop/break out of the iterator.
+                        if (response.StatusCode == HttpStatusCode.NotModified)
+                        {
+                            break;
+                        }                   
+                    }
+                    if(config.PreCreatedDocuments == documentTotal)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:86]:
```suggestion
                    if (config.PreCreatedDocuments == documentTotal) 
```

Comment (ealsur) [reply Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:86]: new line after `}`

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                        continuation = response.ContinuationToken;
+                        //If there are no new changes, stop/break out of the iterator.
+                        if (response.StatusCode == HttpStatusCode.NotModified)
+                        {
+                            break;
+                        }                   
+                    }
+                    if(config.PreCreatedDocuments == documentTotal) 
+                    {
+                        logger.LogInformation("Suceess: The number of new documents match the number of pre-created documents");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:88]:
```suggestion
                        logger.LogInformation($"Success: The number of new documents match the number of pre-created documents {config.PreCreatedDocuments}");
```

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+                    }
+                    //Compare the detected changes vs the prepopulated ones and if they don't match, log a warning with a text showing the 2 values and also log the
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:90]: Remove comments

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+                    }
+                    if(config.PreCreatedDocuments == documentTotal) 
+                    {
+                        logger.LogInformation("Suceess: The number of new documents match the number of pre-created documents");
+                    }
+                    //Compare the detected changes vs the prepopulated ones and if they don't match, log a warning with a text showing the 2 values and also log the 
+                    //value of the last captured Continuation.
+                    else
+                    {
+                        logger.LogError("The prepopulated documents and the new documents don't match.  Preconfigured Docs =" + config.PreCreatedDocuments + ", New Documents = "+ documentTotal);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:94]: Can be simplified with `$`:

```suggestion
                        logger.LogError($"The prepopulated documents and the new documents don't match.  Preconfigured Docs = {config.PreCreatedDocuments}, New Documents {documentTotal}");
```

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+                        logger.LogError(continuation);
+                    }
+                }
+                catch (Exception ex) 
+                {
+                    logger.LogError(ex, "Failure while looping through new documents");
+                    //If there are any Exceptions in the loop (try/catch), log the errors
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:101]: delete comment

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+                        logger.LogError(continuation);
+                    }
+                }
+                catch (Exception ex) 
+                {
+                    logger.LogError(ex, "Failure while looping through new documents");
+                    //If there are any Exceptions in the loop (try/catch), log the errors 
+                }
+            }
+            stopWatch.Stop();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:104]: new line after `}`

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+            [JsonProperty("LeaseToken")]
+            public string LeaseToken { get; set; }
+
+            [JsonProperty("FeedRange")]
+            public Range FeedRange { get; set; }
+
+        }
+
+        internal class Range
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:168]: this class is not used

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+            public bool CreatedContainer;
+        }
+
+        private class SimpleItem
+        {
+            [JsonProperty("id")]
+            public string Id { get; set; }
+        }
+
+        internal class LeaseSchema
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:155]: this class is not used

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+
+        }
+
+        internal class Range
+        {
+            [JsonProperty("Range")]
+            public FeedRange EffectiveRange { get; set; }
+        }
+
+        internal class FeedRange
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:174]: this class is not used

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+            return result;
+        }
+
+        private struct InitializationResult
+        {
+            public bool CreatedDatabase;
+            public bool CreatedContainer;
+        }
+
+        private class SimpleItem
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:149]: This class is not used

**Review by ealsur — CHANGES_REQUESTED** [review]

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task RunAsync(
+            CTLConfig config,
+            CosmosClient cosmosClient,
+            ILogger logger,
+            IMetrics metrics,
+            string loggingContextIdentifier,
+            CancellationToken cancellationToken)
+        {
+            Stopwatch stopWatch = Stopwatch.StartNew();
+            stopWatch.Start();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:45]: Remove the Start, it is not needed.

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+                        if (response.StatusCode == HttpStatusCode.NotModified)
+                        {
+                            break;
+                        }
+                    }
+                    if (config.PreCreatedDocuments == documentTotal)
+                    {
+                        logger.LogInformation($"Suceess: The number of new documents match the number of pre-created documents: {config.PreCreatedDocuments}");
+                    }
+
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:70]: remove empty line

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+                    else
+                    {
+                        logger.LogError($"The prepopulated documents and the new documents don't match.  Preconfigured Docs = {config.PreCreatedDocuments}, New Documents = {documentTotal}");
+                        logger.LogError(continuation);
+                    }
+                }
+
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:77]: remove empty line

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+                        documentTotal += response.Count;
+                        continuation = response.ContinuationToken;
+                        if (response.StatusCode == HttpStatusCode.NotModified)
+                        {
+                            break;
+                        }
+                    }
+                    if (config.PreCreatedDocuments == documentTotal)
+                    {
+                        logger.LogInformation($"Suceess: The number of new documents match the number of pre-created documents: {config.PreCreatedDocuments}");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:68]:
```suggestion
                        logger.LogInformation($"Success: The number of new documents match the number of pre-created documents: {config.PreCreatedDocuments}");
```

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+                    {
+                        FeedResponse<Dictionary<string, string>> response = await changeFeedPull.ReadNextAsync();
+                        documentTotal += response.Count;
+                        continuation = response.ContinuationToken;
+                        if (response.StatusCode == HttpStatusCode.NotModified)
+                        {
+                            break;
+                        }
+                    }
+                    if (config.PreCreatedDocuments == documentTotal)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:66]: new line between the `}` and `if`

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace CosmosCTL
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Net;
+    using System.Diagnostics;
+
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:13]: remove empty lines between usings

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs (hunk 25 above)
Comment (jenglish81) [reply Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs]: Fixed, Thank you.

File: Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+    internal class ChangeFeedPullScenario : ICTLScenario
+    {
+
+        public async Task InitializeAsync(
+            CTLConfig config,
+            CosmosClient cosmosClient,
+            ILogger logger)
+        {
+
+            if (config.PreCreatedDocuments > 0)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Tools/CTL/Scenarios/ChangeFeedPullScenario.cs:27]: You are missing the creation  of the container that you initially had?

**Review by ealsur — APPROVED** [review]
Looks good! Thanks @jenglish81 !

**Comment by jenglish81** [comment]
> Looks good! Thanks @jenglish81 !

Thank you for your help!

**Pull request merged**